### DISCUSSION TITLE
Selecting "Save as new Asset" in image editors opened via Asset fields should replace the originally selected Asset in the field

### DISCUSSION THREAD

**Discussion by mmikkel** [craftcms/cms#8974] [Ideas] [answered]
### Description

It'd be great if, when saving an edited image as a new asset in the image editor _in the context of an Assets field_ (i.e., by double-clicking on a selected asset in an Assets field and opening the image editor from there), Craft would automatically swap out the originally selected asset in the Assets field for the newly created, edited asset.

The current behaviour, where the original file is retained in the field, feels less than intuitive, because I think what _most_ authors want to achieve when clicking "Save as new Asset" in this context is "save my edits to the selected image, but keep a backup of the original file", not "I still want to use the original image in this field, but save my changes to a new file somewhere else".

If selecting the new asset automatically sounds like a bad idea, maybe there could be a prompt when the image editor closes, or perhaps additional actions in the image editor to cover both use cases (i.e. giving the author a choice of keeping the original asset selected in the field or not). I have a pretty strong feeling that auto-selecting the edited asset should be the default action, though.

**Comment by sm9** [comment]
We have a client that has raised a similar point to this in Craft 3.7.12:

> If you edit an image in the CMS (crop etc) and save as a new asset, then I would expect the new edited asset to replace the original asset in that entry. Instead, the new edited asset is saved to Assets, and the old one is retained on the entry. This adds additional steps to the process; you have to manually remove the original asset and then add the asset that you've just edited.

It would be great if this behaviour could be looked at, thanks.

**Reply by thupsi** [reply]
I started a similar [discussion](https://github.com/craftcms/cms/discussions/9687#discussion-35052950)  a while ago but got no response. Maybe I wasn't clear in my description. I hope your suggestion gets more traction!

**Reply by mmikkel** [reply]
Yeah, hope this one receives some attention. In terms of UX, the current behaviour almost feels like a bug (and I've talked with authors who actually thought it was).

**Comment by andris-sevcenko** [comment, marked as answer]
Just added this for Craft 4. (https://github.com/craftcms/cms/commit/6764c9ab5404f1384dd2fac539e3d47740e124be)

**Reply by mmikkel** [reply]
Great, thanks Andris!

**Reply by thupsi** [reply]
Hmm, this does not work for me in 4.0.0-beta.2. Should I file a bug?

**Reply by mmikkel** [reply]
Yeah, doesn't work on my end either – tested on both 4.0.0-beta.1 and 4.0.0-beta.2.

**Reply by brandonkelly** [reply]
Doh sorry! I accidentally reverted this change when adding the new unified element editor stuff. Got it working again for the next release (40cc20ec689eca15bc298741fde0785288b4c1b1).

**Reply by brandonkelly** [reply]
4.0.0-beta.3 is out now with that fix.

**Comment by melbazany** [comment]
Mentioned in another discussion [here](https://github.com/craftcms/cms/discussions/8547#discussioncomment-13390452), but would appreciate having this for clients in Craft 5. It _feels_ like a bug to our clients.

**Reply by mmikkel** [reply]
I just tested it, and as far as I can tell the behaviour described above (which was added in Craft 4.0) still works on Craft 5?

**Reply by melbazany** [reply]
@mmikkel Here's the steps to recreate– _Craft 5.7.10_:

1. Open entry that contains Asset field. Select an asset for the field, and save the entry.
2. Select the "edit" dots -> "Edit asset"
3. In menu drawer, select "Edit Image". Asset edit overlay appears. Select "Crop" and drag handles to crop image accordingly. Click "Save as new asset".
4. The edit asset overlay menu closes, and the original asset file still shows in the menu drawer.
5. Upon saving, there is no change to the newly created asset either. The user needs to manually choose "Replace image" in order to find the newly created asset.

_Expected behavior:_
When clicking "Save as new asset" (3), I would expect the newly cropped image to be populated in the menu drawer, indicating that it has replaced the original asset file.

_Actual behavior:_
When clicking "Save as new asset" (3), the original asset file appears in the menu drawer, indicating that the original asset file is still the selected asset for the field.

Am I missing something?
https://github.com/user-attachments/assets/42d8ac73-e969-4717-b97a-8ff132e94bc7

**Reply by mmikkel** [reply]
@melbazany Aha – so it works when opening the asset slideout (aka the "overlay drawer") and the image editor from there by _double-clicking_ selected assets' thumbnails in asset fields, but not when opening the asset slideout from the Edit action in cards' action menus and etc. 

I wasn't aware of that discrepancy, and I'm not sure if it should be considered a bug, or intended behavior – although I'm leaning towards that the behavior should probably be the same regardless of the method used to open the asset slideout and image editor, and as long as the slideout was opened in the context of an asset field, the newly saved, edited asset should always replace the original asset selected in the field. Maybe open an issue for this?

(Also, note that even when it works, the slideout itself isn't actually reloaded with the "new" asset – the only thing that changes is that the original asset selected in the asset field is swapped out with the newly saved, edited asset – which is what the original FR suggested, but the original FR even pre-dates those slideouts! FWIW, I do agree that ideally, the asset slideout should also be reloaded with the newly saved, edited asset, when "Save as new asset" is selected in the editor view).

**Reply by melbazany** [reply]
Created an issue #17453. Not entirely sure this is a bug either, but I think that less savvy CMS users can easily become confused as to where their new image has gone, when following the described steps.
